language-plutus-core/src/Language/PlutusCore/Clone.hs

```diff
@@ -0,0 +1,28 @@
+module Language.PlutusCore.Clone ( cloneType
+                                 ) where
+
+import           Data.Functor.Foldable
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Quote
+import           Language.PlutusCore.Type
+
+cloneType :: MonadQuote m => Type TyNameWithKind a -> m (Type TyNameWithKind a)
+cloneType (TyFix l (TyNameWithKind (TyName (Name l' n u))) ty) = do
+    u' <- liftQuote freshUnique
+    TyFix l (TyNameWithKind (TyName (Name l' n u'))) <$> cloneType (tyRewrite u u' ty)
+cloneType (TyForall l (TyNameWithKind (TyName (Name l' n u))) k ty) = do
+    u' <- liftQuote freshUnique
+    TyForall l (TyNameWithKind (TyName (Name l' n u'))) k <$> cloneType (tyRewrite u u' ty)
+cloneType (TyLam l (TyNameWithKind (TyName (Name l' n u))) k ty) = do
+    u' <- liftQuote freshUnique
+    TyLam l (TyNameWithKind (TyName (Name l' n u'))) k <$> cloneType (tyRewrite u u' ty)
+cloneType (TyApp l ty ty') = TyApp l <$> cloneType ty <*> cloneType ty'
```

Comment:
May I suggest defining a helper function that matches the `(TyNameWithKind (TyName (Name l' n u)))` and gets the unique? (Currently, the same code is repeated three times.)